### ISSUE TITLE
Enhancement: [no-floating-promises] Provide a built-in SafePromise type

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10666]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises

### Description

Coming over from discussions such as #5844, #7008, #8404, https://github.com/reduxjs/redux-toolkit/issues/4101, https://github.com/fastify/fastify/issues/5498, https://github.com/typescript-eslint/typescript-eslint/issues/9869: one known user pain with `@typescript-eslint/no-floating-promises` is around frameworks/libraries that intentionally make "fire-and-forget" Promises. If a function's calls don't need an `await` by design, it's inconvenient to have the rule report on them by default. 

The [`allowForKnownSafeCalls`](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafecalls) and [`allowForKnownSafePromises`](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises) rule options make this better, but also require user linter configuration. For example, Redux Toolkit added a `SafePromise` type per `allowForKnownSafePromises`. Users can allowlist that type in their ESLint configs. This works but is inconvenient. Most users don't want to deeply hand-edit their ESLint configs for frameworks-specific nuances like this.

We are occasionally asked: can typescript-eslint provide its own `SafePromise` type that frameworks can use, and that `no-floating-promises` would allowlist/ignore by default?

### Fail

```TypeScript
declare function fromFramework(): Promise<void>;

await fromFramework();
// ❌ Lint report from no-floating-promises
```

### Pass

```TypeScript
import { SafePromise } from "@typescript-eslint/types";

declare function fromFramework(): SafePromise<void>;

await fromFramework();
// ✅ No report
```

### Additional Info

I don't see an issue like this in our issue tracker, but know it's been suggested in at least some DMs to me. So I'm surfacing it here on behalf of users. 

💖

**Comment by JoshuaKGoldberg** [comment]
It's been >3 months since we added the `evaluating community engagement` label to this issue. In that time, we have not seen any engagement in favor of this feature request. We're still of the opinion that if something like this exists it cannot come from a single linter team like us -- it'd need to be centralized and/or from a community consensus. 

Closing out as out of scope for this project. If anybody has significant new information for the issue of a built-in "SafePromise" type or similar, such as community consensus on a single type, please do post back in a new issue. We'd be happy to take a look.

For folks who want a no-floating-promises rule that respects a SafePromise type, you're welcome to use the docs in https://typescript-eslint.io/developers to build a rule/plugin that has that. You can always fork the rule from tseslint.

Cheers! 💖
